### DISCUSSION TITLE
Redirect back without query string

### DISCUSSION THREAD

**Discussion by josephmiclaus** [laravel/framework#41612] [Ideas] [answered]
Sometimes you want to redirect back, however without the query string.

e.g. for the URL: `website.dev/foo?bar=1` submit form and then `return back();` in the controller redirects to that URL (including the query string).

However I would like to redirect to `website.dev/foo` (without the query string).

One could instead `return redirect()->route('something');` or `return redirect('/something');`, however not every route has a route name and you might want to return to the previous route without knowing which one it was.

Is there already an elegant way to achieve this? (I did some research, but couldn't find anything useful.)

If there isn't, could we maybe implement one of the following?
```php
return back(ignoreQuery: true);
```
``` php
return redirect(url()->previous(ignoreQuery: true));
```
```php
return redirect(url()->previousWithoutQuery());
```

If I'm not mistaken, the first and second possibilities might introduce breaking changes.

The third possibility would be easy to implement in `UrlGenerator` (plus related changes to the `url()` helper and `URL` facade):

```php
/**
 * Get the URL for the previous request without the query string.
 * 
 * @param  mixed  $fallback
 * @return string
 */
public function previousWithoutQuery($fallback = false)
{
    return rtrim(preg_replace('/\?.*/', '', $this->previous($fallback)), '/');
}
```
Not sure if `$fallback` would be needed, but it's possible.

---

Then in turn, would make the following possible:

```php
/**
 * Get the previous path info for the request.
 * 
 * @param  mixed  $fallback
 * @return string
 */
public function previousPath($fallback = false)
{
    return str_replace($this->to('/'), '', $this->previousWithoutQuery($fallback));
}
```

Enabling:
```php
return redirect(url()->previousPath());
```

I'd be happy to do a PR.

**Comment by ralphjsmit** [comment]
I think you can do this:

```php
// Redirect to same url with "ref" query string. Also accepts an array with keys.
return redirect()->away($request->fullUrlWithoutQuery('ref'));
```

**Reply by josephmiclaus** [reply]
@ralphjsmit I knew it should be possible somehow!

But in this case you have to know the parameters such as "ref" from the query string, right? What if you don't know all specified parameters?

Btw. `redirect(url()->previousPath());` has been merged and should be usable now.

**Reply by ralphjsmit** [reply]
Hmm yes, that's also true indeed. Probably you could figure them out with the $request object, but yes, it's indeed for removing specific parameters.

**Comment by Kristories** [comment]
@miclaus I think, the simple way is like this:

```
$back = strtok(redirect()->back()->getTargetUrl(), '?');

return redirect($back);
```
- Get `RedirectResponse` with `redirect()->back()` and get target URL via `getTargetUrl()`
- Remove query string with `strtok()`
- Then redirect back to the modified url

**Comment by josephmiclaus** [comment, marked as answer]
@Kristories since PR #41661 you can do this:

```php
return redirect(url()->previousPath());
```
